kep-135 add startup policy api (#136)
kubernetes-sigs/lws: keps/135-startup-policy/README.md

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### API
```go
type StartupPolicyType string

const (
	// LeaderFirst will create worker statefulset after leader statefulset is ready.

> 	LeaderFirstStartupPolicy StartupPolicyType = "LeaderFirst"

Author: sry for missing that. 
Addressed.

	AnyOrderStartupPolicy StartupPolicyType = "AnyOrder"
)

type LeaderWorkerSetSpec struct {
	// StartupPolicy determines the startup order of the leader statefulSet and worker statefulSets.
	// +kubebuilder:default=AnyOrder
	// +kubebuilder:validation:Enum={AnyOrder,LeaderFirst}
	StartupPolicy StartupPolicyType `json:"startupPolicy"`
}
```

### Implementation
We don't need to change the logic when `spec.startupPolicy=AnyOrder`. We only need to implement the logic when `spec.startupPolicy=LeaderFirst`.

#### lws webhook
We need to set default value of `spec.startupPolicy` to `AnyOrder` in the lws webhook.

#### Pod controller
Pod controller is responsible for creating worker statefulset. When using `LeaderFirst` strategy, we only need to care about when to create worker statefulset.
Here is the reconcile logic of creating a new lws with `spec.startupPolicy=LeaderFirst`:
1. A new leaderworkerset was created.
2. The leader statefulset was created by lws controller but leader pod is not ready yet.
3. Pod controller will watch the leader pod `Update` events and get the leader statefulset latest status to determine if we should create the worker statefulsets.

>     - Do nothing if `leader.status.readyReplicas != leader.status.replicas`, because the leader statefulset is not ready at this time.

Author: Thanks for catching this and yes, you're right. I'll update the logic later.

    - [Unit Tests](#unit-tests)
    - [Integration tests](#integration-tests)
    - [e2e tests](#e2e-tests)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

> This KEP aims to add `spec.startupPolicy` filed to LeaderWorkerSet's API, to indicate the startup order of leader statefulset and worker statefulsets.

Reviewer: ```suggestion
This KEP aims to add `spec.startupPolicy` filed to LeaderWorkerSet's API, to indicate the startup order of the leader pod and its worker statefulset in the group.
```

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

> Some frameworks require the leader statefulset in LeaderWorkerSet to be ready (like ray head) before the worker statefulsets are started so they can connect to it.

Reviewer 2: ```suggestion
Some frameworks require the leader in LeaderWorkerSet to be ready (like ray head) before the reset of the workers are started so they can connect to it.
```

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

> - Add a new API `WaitForLeaderReady` to indicate if we need to create worker statefulsets until the leader statefulset is ready.

Reviewer: ```suggestion
- Add a way to allow specifying different startup policies. For ex, to indicate if we need to create the worker statefulset until its leader pod is ready.
```

> - Add a way to allow specifying different startup policies. For ex, to indicate if we need to create the worker statefulset until its leader pod is ready.

Reviewer 2: ```suggestion
- Add a way to allow specifying different startup policies. For example, to indicate that the leader must be ready before the rest of the workers are created.
```

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

> We propose a new field `spec.startupPolicy` to the LeaderWorkerSet API. This field is defaulted to `Default` which means that lws controller will create worker statefulsets immediately after leader statefulset created. It doesn't guarantee the order of readiness for leader statefulset and worker statefulsets.

Reviewer: ```suggestion
We propose a new field `spec.startupPolicy` to the LeaderWorkerSet API. This field is defaulted to `Default` which means that lws controller will create the worker statefulset immediately after the leader pod is created. It doesn't guarantee the order of readiness for leader statefulset and worker statefulsets.
```

> We propose a new field `spec.startupPolicy` to the LeaderWorkerSet API. This field is defaulted to `Default` which means that lws controller will create the worker statefulset immediately after the leader pod is created. It doesn't guarantee the order of readiness for leader statefulset and worker statefulsets.

Reviewer 2: I suggest two values: `AnyOrder` which is the default, and `LeaderFirst` which indicates that the leader must startup first before the rest of the workers.

> Another option for this field is `WaitForLeaderReady`. It will delay creating the worker statefulsets until the leader statefulset is ready.

Reviewer: ```suggestion
Another option for this field is `WaitForLeaderReady`. It will delay creating the worker statefulset until its leader pod is ready.
```

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

> #### Story 1

Reviewer 2: can you add the Ray story here pls?

Author: sry, I'm not familiar with Ray story. Can you briefly introduce the Ray story and then let me add it into the kep? Thanks.

Reviewer 2: "Ray runs a head and a set of workers, and we want the Ray head to run by the leader and be ready first before the workers are started since the workers connect back to the head."

> #### Story 2

Reviewer 2: remove this header pls

Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### API
```go
type StartupPolicyType string

const (
	// WaitForLeaderReady will create worker statefulset after leader statefulset is ready.
	WaitForLeaderReady StartupPolicyType = "WaitForLeaderReady"

	// Default will create worker statefulsets immediately while leader statefulset is created.
	DefaultStartupPolicy StartupPolicyType = "Default"
)

type LeaderWorkerSetSpec struct {
	// StartupPolicy determines the startup order of the leader statefulSet and worker statefulSets.
	// +kubebuilder:default=Default
	// +kubebuilder:validation:Enum={Default,WaitForLeaderReady}
	StartupPolicy StartupPolicyType `json:"startupPolicy"`
}
```

### Implementation

> We don't need to change the logic when `spec.startupPolicy=Default`. We only need to implement the logic when `spec.startupPolicy=WaitForLeaderReady`.

Reviewer: what would be the defaulting logic in webhook

<!-- #### lsw controller
lws controller will set worker statefulset's replicas=0 before all leader statefulset pods are ready. -->

#### Pod controller
Pod controller is responsible for creating worker statefulset. When using `WaitForLeaderReady` strategy, we only need to care about when to create worker statefulset.
Here is the reconcile logic of creating a new lws with `spec.startupPolicy=WaitForLeaderReady`:
1. A new leaderworkerset was created.
2. The leader statefulset was created by lws controller but leader pod is not ready yet.

> 3. Pod controller watch the leader pod created/updated event and begin to constructs the worker statefulset configuration. Firstly, it will get the leader statefulset latest status to determine if we should create the worker statefulset pods.

Reviewer: we could construct the worker sts configuration after checking the leader pod is ready

Reviewer 2: +1, create the workers sts only after the leader is ready

- <package>: <date> - <current test coverage>
The data can be easily read from:
https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

All core changes must be covered by unit tests, in both API and pod controller.
- API: API validation.
- lws controller
- Pod controller

- `<package>`: `<date>` - `<test coverage>`

##### Integration tests

<!--
Integration tests are contained in k8s.io/kubernetes/test/integration.
Integration tests allow control of the configuration parameters used to start the binaries under test.
This is different from e2e tests which do not allow configuration of parameters.
Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

> 

Author: Haven't figured out a good way to write integration tests.

Reviewer 3: We can control whether Pod is ready or not in integration tests because we have no kubelet here.

- <test>: <link to test coverage>

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

> Ddeploy lws with `spec.startupPolicy=WaitForLeaderReady` set.

Reviewer: ```suggestion
Deploy lws with `spec.startupPolicy=WaitForLeaderReady` set.
```
